File: Telephone/ActiveCallViewController.m
//
//  ActiveCallViewController.m
//  Telephone
//
//  Copyright © 2008-2016 Alexey Kuznetsov
//  Copyright © 2016-2022 64 Characters
//
//  Telephone is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//
//  Telephone is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//

#import "ActiveCallViewController.h"

#import "AKNSWindow+Resizing.h"
#import "AKSIPCall.h"

#import "CallController.h"
#import "CallTransferController.h"
#import "EndedCallViewController.h"


@interface ActiveCallViewController () <NSMenuItemValidation>

@property(nonatomic, getter=isShowingProgress) BOOL showingProgress;

@property(nonatomic, weak) IBOutlet NSTextField *displayedNameField;
@property(nonatomic, weak) IBOutlet NSTextField *statusField;

@property(nonatomic) IBOutlet NSProgressIndicator *callProgressIndicator;
@property(nonatomic) IBOutlet NSButton *hangUpButton;
@property(nonatomic) NSButton *muteButton;
@property(nonatomic) NSButton *holdButton;
@property(nonatomic) NSButton *transferButton;

@end

@implementation ActiveCallViewController

- (instancetype)initWithNibName:(NSString *)nibName callController:(CallController *)callController {
    self = [super initWithNibName:nibName bundle:nil];
    
    if (self != nil) {
        _enteredDTMF = [[NSMutableString alloc] init];
        [self setCallController:callController];
    }
    return self;
}

- (instancetype)init {
    NSString *reason = @"Initialize ActiveCallViewController with initWithCallController:";
    @throw [NSException exceptionWithName:@"AKBadInitCall" reason:reason userInfo:nil];
    return nil;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self configureCallControls];
    [self updateCallControls];
}

- (void)configureCallControls {
    self.hangUpButton.hidden = NO;
    self.hangUpButton.image = [NSImage imageWithSystemSymbolName:@"phone.down.fill"
                                        accessibilityDescription:NSLocalizedString(@"End Call", @"End call button.")];
    self.hangUpButton.imagePosition = NSImageOnly;
    self.hangUpButton.toolTip = NSLocalizedString(@"End Call", @"End call button.");

    for (NSLayoutConstraint *constraint in self.hangUpButton.constraints) {
        if (constraint.firstAttribute == NSLayoutAttributeWidth ||
            constraint.firstAttribute == NSLayoutAttributeHeight) {
            constraint.constant = 28;
        }
    }

    NSLayoutConstraint *progressTrailingConstraint = nil;
    for (NSLayoutConstraint *constraint in self.view.constraints) {
        if (constraint.firstItem == self.view &&
            constraint.firstAttribute == NSLayoutAttributeTrailing &&
            constraint.secondItem == self.callProgressIndicator &&
            constraint.secondAttribute == NSLayoutAttributeTrailing) {
            progressTrailingConstraint = constraint;
            break;
        }
    }
    progressTrailingConstraint.active = NO;

    [self.callProgressIndicator.trailingAnchor
        constraintEqualToAnchor:self.hangUpButton.leadingAnchor
                       constant:-8].active = YES;

    self.muteButton = [NSButton buttonWithImage:[NSImage imageWithSystemSymbolName:@"mic.fill"
                                                          accessibilityDescription:nil]
                                         target:self
                                         action:@selector(toggleMicrophoneMute:)];
    self.holdButton = [NSButton buttonWithImage:[NSImage imageWithSystemSymbolName:@"pause.fill"
                                                          accessibilityDescription:nil]
                                         target:self
                                         action:@selector(toggleCallHold:)];
    self.transferButton = [NSButton buttonWithImage:[NSImage imageWithSystemSymbolName:@"arrow.right"
                                                              accessibilityDescription:nil]
                                             target:self
                                             action:@selector(showCallTransferSheet:)];

    for (NSButton *button in @[self.muteButton, self.holdButton, self.transferButton]) {
        button.translatesAutoresizingMaskIntoConstraints = NO;
        button.bezelStyle = NSBezelStyleTexturedRounded;
        button.imagePosition = NSImageOnly;
        [self.view addSubview:button];
    }
    self.muteButton.buttonType = NSButtonTypeToggle;
    self.holdButton.buttonType = NSButtonTypeToggle;

    [NSLayoutConstraint activateConstraints:@[
        [self.transferButton.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor constant:-20],
        [self.transferButton.centerYAnchor constraintEqualToAnchor:self.statusField.centerYAnchor],
        [self.transferButton.widthAnchor constraintEqualToConstant:28],
        [self.transferButton.heightAnchor constraintEqualToConstant:28],
        [self.holdButton.trailingAnchor constraintEqualToAnchor:self.transferButton.leadingAnchor constant:-8],
        [self.holdButton.centerYAnchor constraintEqualToAnchor:self.statusField.centerYAnchor],
        [self.holdButton.widthAnchor constraintEqualToConstant:28],
        [self.holdButton.heightAnchor constraintEqualToConstant:28],
        [self.muteButton.trailingAnchor constraintEqualToAnchor:self.holdButton.leadingAnchor constant:-8],
        [self.muteButton.centerYAnchor constraintEqualToAnchor:self.statusField.centerYAnchor],
        [self.muteButton.widthAnchor constraintEqualToConstant:28],
        [self.muteButton.heightAnchor constraintEqualToConstant:28],
        [self.muteButton.leadingAnchor constraintGreaterThanOrEqualToAnchor:self.statusField.trailingAnchor constant:8]
    ]];
}

- (void)removeObservations {
    [[self displayedNameField] unbind:NSValueBinding];
    [[self statusField] unbind:NSValueBinding];
 }

- (IBAction)hangUpCall:(id)sender {
    [[self callController] hangUpCall];
}

- (IBAction)toggleCallHold:(id)sender {
    BOOL held = self.callController.call.isOnLocalHold;
    [self.callController setCallHeld:!held];
    [self updateCallControls];
}

- (IBAction)toggleMicrophoneMute:(id)sender {
    BOOL muted = self.callController.call.isMicrophoneMuted;
    [self.callController setMicrophoneMuted:!muted];
    [self updateCallControls];
}

- (IBAction)showCallTransferSheet:(id)sender {
    if (![[self callController] isCallOnHold]) {
        [[self callController] setCallHeld:YES];
    }
    
    CallTransferController *callTransferController = [[self callController] callTransferController];

    [[[self callController] window] beginSheet:[callTransferController window] completionHandler:nil];
}

- (void)startCallTimer {
    if ([self callTimer] != nil && [[self callTimer] isValid]) {
        return;
    }
    
    [self setCallTimer:
     [NSTimer scheduledTimerWithTimeInterval:0.2
                                      target:self
                                    selector:@selector(callTimerTick:)
                                    userInfo:nil
                                     repeats:YES]];
}

- (void)stopCallTimer {
    if ([self callTimer] != nil) {
        [[self callTimer] invalidate];
        [self setCallTimer:nil];
    }
}

- (void)callTimerTick:(NSTimer *)theTimer {
    NSTimeInterval now = [NSDate timeIntervalSinceReferenceDate];
    NSInteger seconds = (NSInteger)(now - ([[self callController] callStartTime]));
    
    if (seconds < 3600) {
        [[self callController] setStatus:[NSString stringWithFormat:@"%02ld:%02ld",
                                          (seconds / 60) % 60, seconds % 60]];
    } else {
        [[self callController]
         setStatus:[NSString stringWithFormat:@"%02ld:%02ld:%02ld",
                    (seconds / 3600) % 24, (seconds / 60) % 60, seconds % 60]];
    }
}

- (void)showProgress {
    if (!self.isShowingProgress) {
        [self.callProgressIndicator startAnimation:self];
        self.showingProgress = YES;
    }
    self.callProgressIndicator.hidden = NO;
    self.hangUpButton.hidden = NO;
    [self updateCallControls];
}

- (void)showHangUp {
    if (self.isShowingProgress) {
        [self.callProgressIndicator stopAnimation:nil];
        self.showingProgress = NO;
    }
    self.callProgressIndicator.hidden = YES;
    self.hangUpButton.hidden = NO;
    [self updateCallControls];
}

- (void)allowHangUp {
    self.hangUpButton.enabled = YES;
}

- (void)disallowHangUp {
    self.hangUpButton.enabled = NO;
}

- (void)updateCallControls {
    AKSIPCall *call = self.callController.call;
    BOOL confirmed = call.state == kAKSIPCallConfirmedState;

    BOOL muted = call.isMicrophoneMuted;
    self.muteButton.enabled = confirmed;
    self.muteButton.state = muted ? NSControlStateValueOn : NSControlStateValueOff;
    self.muteButton.image = [NSImage imageWithSystemSymbolName:(muted ? @"mic.slash.fill" : @"mic.fill")
                                      accessibilityDescription:nil];
    NSString *muteTitle = muted
        ? NSLocalizedString(@"Unmute", @"Unmute. Call menu item.")
        : NSLocalizedString(@"Mute", @"Mute. Call menu item.");
    self.muteButton.toolTip = muteTitle;
    self.muteButton.accessibilityLabel = muteTitle;

    BOOL held = call.isOnLocalHold;
    self.holdButton.enabled = confirmed && !call.isOnRemoteHold;
    self.holdButton.state = held ? NSControlStateValueOn : NSControlStateValueOff;
    self.holdButton.image = [NSImage imageWithSystemSymbolName:(held ? @"play.fill" : @"pause.fill")
                                      accessibilityDescription:nil];
    NSString *holdTitle = held
        ? NSLocalizedString(@"Resume", @"Resume. Call menu item.")
        : NSLocalizedString(@"Hold", @"Hold. Call menu item.");
    self.holdButton.toolTip = holdTitle;
    self.holdButton.accessibilityLabel = holdTitle;

    BOOL transferEnabled = confirmed && !call.isOnRemoteHold;
    NSString *transferTitle = NSLocalizedString(@"Transfer", @"Transfer. Call menu item.");
    self.transferButton.enabled = transferEnabled;
    self.transferButton.toolTip = transferTitle;
    self.transferButton.accessibilityLabel = transferTitle;
}


#pragma mark -
#pragma mark AKActiveCallViewDelegate protocol

- (void)activeCallView:(AKActiveCallView *)sender didReceiveText:(NSString *)aString {
    NSCharacterSet *DTMFCharacterSet = [NSCharacterSet characterSetWithCharactersInString:@"0123456789*#abcdrABCDR"];
    
    BOOL isDTMFValid = YES;
    for (NSUInteger i = 0; i < [aString length]; ++i) {
        unichar digit = [aString characterAtIndex:i];
        if (![DTMFCharacterSet characterIsMember:digit]) {
            isDTMFValid = NO;
            break;
        }
    }
    
    if (isDTMFValid) {
        if ([[self enteredDTMF] length] == 0) {
            [[self enteredDTMF] appendString:aString];
            [[[self view] window] setTitle:[[self callController] displayedName]];
            
            if ([[self displayedNameField] lineBreakMode]!= NSLineBreakByTruncatingHead) {
                [[self displayedNameField] setLineBreakMode:NSLineBreakByTruncatingHead];
                [[[[self callController] endedCallViewController] displayedNameField] setSelectable:YES];
            }
            
            [[self callController] setDisplayedName:aString];
            
        } else {
            [[self enteredDTMF] appendString:aString];
            [[self callController] setDisplayedName:[self enteredDTMF]];
        }
        
        [[[self callController] call] sendDTMFDigits:aString];
    }
}


#pragma mark -
#pragma mark NSMenuItemValidation protocol

- (BOOL)validateMenuItem:(NSMenuItem *)menuItem {
    if ([menuItem action] == @selector(toggleMicrophoneMute:)) {
        if ([[[self callController] call] isMicrophoneMuted]) {
            [menuItem setTitle:NSLocalizedString(@"Unmute", @"Unmute. Call menu item.")];
        } else {
            [menuItem setTitle:NSLocalizedString(@"Mute", @"Mute. Call menu item.")];
        }
        
        if ([[[self callController] call] state] == kAKSIPCallConfirmedState) {
            return YES;
        } else {
            return NO;
        }
        
    } else if ([menuItem action] == @selector(toggleCallHold:)) {
        if ([[[self callController] call] state] == kAKSIPCallConfirmedState &&
            [[[self callController] call] isOnLocalHold]) {
            [menuItem setTitle:NSLocalizedString(@"Resume", @"Resume. Call menu item.")];
        } else {
            [menuItem setTitle:NSLocalizedString(@"Hold", @"Hold. Call menu item.")];
        }
        
        if ([[[self callController] call] state] == kAKSIPCallConfirmedState &&
            ![[[self callController] call] isOnRemoteHold]) {
            
            return YES;
            
        } else {
            return NO;
        }
        
    } else if ([menuItem action] == @selector(showCallTransferSheet:)) {
        if ([[[self callController] call] state] == kAKSIPCallConfirmedState &&
            ![[[self callController] call] isOnRemoteHold]) {
            
            return YES;
            
        } else {
            return NO;
        }
        
    } else if ([menuItem action] == @selector(hangUpCall:)) {
        [menuItem setTitle:NSLocalizedString(@"End Call", @"End Call. Call menu item.")];

        return self.hangUpButton.isEnabled;
    }
    
    return YES;
}

@end

File: Telephone/ActiveCallViewController.swift
//
//  ActiveCallViewController.swift
//  Telephone
//

import AppKit
import Observation
import SwiftUI

@MainActor
@Observable
fileprivate final class ActiveCallModel {
    var displayedName = ""
    var status = ""
    var showsProgress = false
    var hangUpEnabled = true
    var muteEnabled = false
    var muted = false
    var holdEnabled = false
    var held = false
    var transferEnabled = false
    var transferActionEnabled = false
    var cancelEnabled = true
    var usesDTMFDisplay = false
}

@MainActor
@objcMembers
class ActiveCallViewController: NSViewController, NSMenuItemValidation {
    weak var callController: CallController?
    var callTimer: Foundation.Timer?
    var enteredDTMF = NSMutableString()

    fileprivate let model = ActiveCallModel()
    private var displayedNameObservation: NSKeyValueObservation?
    private var statusObservation: NSKeyValueObservation?

    @objc(initWithNibName:callController:)
    init(nibName: String, callController: CallController) {
        self.callController = callController
        super.init(nibName: nil, bundle: nil)
        startObserving()
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    fileprivate var isTransferPresentation: Bool {
        self is ActiveCallTransferViewController
    }

    override func loadView() {
        let root = ActiveCallView(
            model: model,
            isTransfer: isTransferPresentation,
            hangUp: { [weak self] in self?.hangUpCall(nil) },
            toggleMute: { [weak self] in self?.toggleMicrophoneMute(nil) },
            toggleHold: { [weak self] in self?.toggleCallHold(nil) },
            showTransfer: { [weak self] in self?.showCallTransferSheet(nil) },
            cancelTransfer: { [weak self] in
                (self as? ActiveCallTransferViewController)?.cancelTransfer(nil)
            },
            completeTransfer: { [weak self] in
                (self as? ActiveCallTransferViewController)?.transferCall(nil)
            }
        )

        let hostingView = DTMFCapturingHostingView(rootView: root)
        hostingView.onText = { [weak self] text in
            self?.handleDTMF(text)
        }
        view = hostingView
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        updateCallControls()
    }

    func removeObservations() {
        displayedNameObservation?.invalidate()
        statusObservation?.invalidate()
        displayedNameObservation = nil
        statusObservation = nil
    }

    @IBAction func hangUpCall(_ sender: Any?) {
        callController?.hangUpCall()
    }

    @IBAction func toggleCallHold(_ sender: Any?) {
        guard let call = callController?.call else { return }
        callController?.setCallHeld(!call.isOnLocalHold)
        updateCallControls()
    }

    @IBAction func toggleMicrophoneMute(_ sender: Any?) {
        guard let call = callController?.call else { return }
        callController?.setMicrophoneMuted(!call.isMicrophoneMuted)
        updateCallControls()
    }

    @IBAction func showCallTransferSheet(_ sender: Any?) {
        guard let callController else { return }

        if !callController.isCallOnHold {
            callController.setCallHeld(true)
        }

        let transferController = callController.callTransferController
        guard
            let transferWindow = transferController?.window,
            let parentWindow = callController.window
        else {
            return
        }
        parentWindow.beginSheet(transferWindow)
    }

    func startCallTimer() {
        guard callTimer?.isValid != true else { return }

        callTimer = Foundation.Timer.scheduledTimer(
            timeInterval: 0.2,
            target: self,
            selector: #selector(callTimerTick(_:)),
            userInfo: nil,
            repeats: true
        )
    }

    func stopCallTimer() {
        callTimer?.invalidate()
        callTimer = nil
    }

    @objc func callTimerTick(_ timer: Foundation.Timer) {
        guard let callController else { return }

        let seconds = Int(Date.timeIntervalSinceReferenceDate - callController.callStartTime)
        if seconds < 3600 {
            callController.status = String(
                format: "%02d:%02d",
                (seconds / 60) % 60,
                seconds % 60
            )
        } else {
            callController.status = String(
                format: "%02d:%02d:%02d",
                (seconds / 3600) % 24,
                (seconds / 60) % 60,
                seconds % 60
            )
        }
    }

    func showProgress() {
        model.showsProgress = true
        updateCallControls()
    }

    func showHangUp() {
        model.showsProgress = false
        updateCallControls()
    }

    func allowHangUp() {
        model.hangUpEnabled = true
    }

    func disallowHangUp() {
        model.hangUpEnabled = false
    }

    func updateCallControls() {
        guard let call = callController?.call else {
            model.muteEnabled = false
            model.holdEnabled = false
            model.transferEnabled = false
            return
        }

        let confirmed = call.isConfirmed
        model.muted = call.isMicrophoneMuted
        model.muteEnabled = confirmed
        model.held = call.isOnLocalHold
        model.holdEnabled = confirmed && !call.isOnRemoteHold
        model.transferEnabled = confirmed && !call.isOnRemoteHold
    }

    func validateMenuItem(_ menuItem: NSMenuItem) -> Bool {
        switch menuItem.action {
        case #selector(toggleMicrophoneMute(_:)):
            menuItem.title = model.muted
                ? NSLocalizedString("Unmute", comment: "Unmute. Call menu item.")
                : NSLocalizedString("Mute", comment: "Mute. Call menu item.")
            return model.muteEnabled

        case #selector(toggleCallHold(_:)):
            menuItem.title = model.held
                ? NSLocalizedString("Resume", comment: "Resume. Call menu item.")
                : NSLocalizedString("Hold", comment: "Hold. Call menu item.")
            return model.holdEnabled

        case #selector(showCallTransferSheet(_:)):
            return model.transferEnabled

        case #selector(hangUpCall(_:)):
            menuItem.title = NSLocalizedString("End Call", comment: "End Call. Call menu item.")
            return model.hangUpEnabled

        default:
            return true
        }
    }

    private func startObserving() {
        guard let callController else { return }

        displayedNameObservation = callController.observe(
            \.displayedName,
            options: [.initial, .new]
        ) { [weak self] controller, _ in
            let value = controller.displayedName ?? ""
            Task { @MainActor [weak self] in
                self?.model.displayedName = value
            }
        }

        statusObservation = callController.observe(
            \.status,
            options: [.initial, .new]
        ) { [weak self] controller, _ in
            let value = controller.status ?? ""
            Task { @MainActor [weak self] in
                self?.model.status = value
            }
        }
    }

    private func handleDTMF(_ text: String) {
        let allowed = CharacterSet(charactersIn: "0123456789*#abcdrABCDR")
        guard text.unicodeScalars.allSatisfy(allowed.contains) else { return }
        guard let callController, let call = callController.call else { return }

        if enteredDTMF.length == 0 {
            view.window?.title = callController.displayedName
            model.usesDTMFDisplay = true
        }

        enteredDTMF.append(text)
        callController.displayedName = enteredDTMF as String
        call.sendDTMFDigits(text)
    }
}

@MainActor
@objcMembers
final class ActiveCallTransferViewController: ActiveCallViewController {
    private var waitingForHold = false

    override var representedObject: Any? {
        didSet {
            waitingForHold = false
        }
    }

    @IBAction func transferCall(_ sender: Any?) {
        guard
            let callController,
            let transferController = callController as? CallTransferController
        else {
            return
        }

        if callController.isCallOnHold {
            transferController.transferCall()
        } else {
            callController.toggleCallHold()
            disallowTransfer()
            waitingForHold = true
        }
    }

    func callDidHold() {
        guard
            waitingForHold,
            let transferController = callController as? CallTransferController
        else {
            return
        }

        transferController.transferCall()
        waitingForHold = false
    }

    func allowTransfer() {
        model.transferActionEnabled = true
    }

    func disallowTransfer() {
        model.transferActionEnabled = false
    }

    @IBAction func cancelTransfer(_ sender: Any?) {
        guard let transferController = callController as? CallTransferController else { return }
        transferController.showInitialState(sender)
    }

    override func showCallTransferSheet(_ sender: Any?) {
        // A transfer target cannot open another transfer sheet.
    }

    override func allowHangUp() {
        super.allowHangUp()
        model.cancelEnabled = true
    }

    override func disallowHangUp() {
        super.disallowHangUp()
        model.cancelEnabled = false
    }

    override func validateMenuItem(_ menuItem: NSMenuItem) -> Bool {
        if menuItem.action == #selector(showCallTransferSheet(_:)) {
            return false
        }
        return super.validateMenuItem(menuItem)
    }
}

private struct ActiveCallView: View {
    @Bindable var model: ActiveCallModel

    let isTransfer: Bool
    let hangUp: () -> Void
    let toggleMute: () -> Void
    let toggleHold: () -> Void
    let showTransfer: () -> Void
    let cancelTransfer: () -> Void
    let completeTransfer: () -> Void

    var body: some View {
        if isTransfer {
            transferBody
                .frame(width: 320, height: 115)
        } else {
            regularBody
                .frame(width: 300, height: 84)
        }
    }

    private var regularBody: some View {
        VStack(alignment: .leading, spacing: 4) {
            HStack(spacing: 8) {
                Text(model.displayedName)
                    .font(.headline)
                    .lineLimit(1)
                    .truncationMode(model.usesDTMFDisplay ? .head : .tail)

                Spacer(minLength: 8)

                if model.showsProgress {
                    ProgressView()
                        .controlSize(.small)
                }

                Button(action: hangUp) {
                    Image(systemName: "phone.down.fill")
                }
                .buttonStyle(.bordered)
                .disabled(!model.hangUpEnabled)
                .help(NSLocalizedString("End Call", comment: "End call button."))
            }

            HStack(spacing: 8) {
                Text(model.status)
                    .foregroundStyle(.secondary)
                    .lineLimit(1)

                Spacer(minLength: 8)

                CallControlButton(
                    systemImage: model.muted ? "mic.slash.fill" : "mic.fill",
                    help: model.muted
                        ? NSLocalizedString("Unmute", comment: "Unmute. Call menu item.")
                        : NSLocalizedString("Mute", comment: "Mute. Call menu item."),
                    isEnabled: model.muteEnabled,
                    action: toggleMute
                )

                CallControlButton(
                    systemImage: model.held ? "play.fill" : "pause.fill",
                    help: model.held
                        ? NSLocalizedString("Resume", comment: "Resume. Call menu item.")
                        : NSLocalizedString("Hold", comment: "Hold. Call menu item."),
                    isEnabled: model.holdEnabled,
                    action: toggleHold
                )

                CallControlButton(
                    systemImage: "arrow.right",
                    help: NSLocalizedString("Transfer", comment: "Transfer. Call menu item."),
                    isEnabled: model.transferEnabled,
                    action: showTransfer
                )
            }
        }
        .padding(.horizontal, 20)
        .padding(.vertical, 10)
    }

    private var transferBody: some View {
        VStack(alignment: .leading, spacing: 4) {
            HStack(spacing: 8) {
                VStack(alignment: .leading, spacing: 4) {
                    Text(model.displayedName)
                        .font(.headline)
                        .lineLimit(1)
                        .truncationMode(model.usesDTMFDisplay ? .head : .tail)

                    Text(model.status)
                        .font(.caption)
                        .foregroundStyle(.secondary)
                        .lineLimit(1)
                }

                Spacer(minLength: 8)

                if model.showsProgress {
                    ProgressView()
                        .controlSize(.small)
                }
            }

            Spacer(minLength: 4)

            HStack {
                Spacer()

                Button(
                    NSLocalizedString("Cancel", comment: "Cancel button."),
                    action: cancelTransfer
                )
                .keyboardShortcut(.cancelAction)
                .disabled(!model.cancelEnabled)

                Button(
                    NSLocalizedString("Transfer", comment: "Transfer call button."),
                    action: completeTransfer
                )
                .keyboardShortcut(.defaultAction)
                .disabled(!model.transferActionEnabled)
            }
        }
        .padding(14)
    }
}

private struct CallControlButton: View {
    let systemImage: String
    let help: String
    let isEnabled: Bool
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            Image(systemName: systemImage)
                .frame(width: 16, height: 16)
        }
        .buttonStyle(.bordered)
        .disabled(!isEnabled)
        .help(help)
        .accessibilityLabel(help)
    }
}

private final class DTMFCapturingHostingView<Content: View>: NSHostingView<Content> {
    var onText: ((String) -> Void)?

    override var acceptsFirstResponder: Bool {
        true
    }

    override func keyDown(with event: NSEvent) {
        guard
            !event.isARepeat,
            let characters = event.characters,
            let firstScalar = characters.unicodeScalars.first
        else {
            super.keyDown(with: event)
            return
        }

        let allowed = CharacterSet(charactersIn: "0123456789*#abcdrABCDR")
        guard allowed.contains(firstScalar) else {
            super.keyDown(with: event)
            return
        }

        onText?(characters)
    }
}
